core/src/main/scala/io/github/gaelrenoux/tranzactio/ConnectionSource.scala

```diff
@@ -39,6 +46,16 @@ object ConnectionSource {
       }
     }
 
+    def runTransactionStream[R, E, A](task: Connection => ZStream[R, E, A], commitOnFailure: => Boolean = false)
+       (implicit errorStrategies: ErrorStrategiesRef, trace: Trace): ZStream[R, Either[DbException, E], A] = {
+
+      ZStream.acquireReleaseWith(openConnection.tap(c => setAutoCommit(c, autoCommit = false)).mapError(Left(_)))
+        { c => commitConnection(c).tapEither(_ => closeConnection(c)).orDie }
+        .flatMap(c => task(c).mapError(Right(_))
+          .tapError(_ => (if (commitOnFailure) commitConnection(c) else rollbackConnection(c)).mapError(Left(_))))
+        .catchSomeCause { case Cause.Die(error: DbException, _) => ZStream.fail(Left(error)) }
```

Comment:
Why do you want to catch the `Die` here? As far as I can see, at this point, the only reason to have the stream die with a `DbException` is if the connection could not be closed.

If that happens, your whole application is pretty much screwed. There's nothing you can do to recover from that apart from stopping the process and restarting. If there's a very corner case when someone might want to catch that and do something, they can catch the `Die` themselves, but that shouldn't be the default behavior.

Reply:
It was a bug in ZIO framework were its not worked correctly with the resource, 
I needed to play with it til I realised it, I also made the Connection to a private layer which could be used in both ZIO and ZStream

```
    private def connectionLayer(commitOnFailure: => Boolean = false)(implicit
        errorStrategies: ErrorStrategiesRef,
        trace: Trace
    ): ZLayer[Any, DbException, Connection] =
      ZLayer
        .scoped {
          ZIO.acquireRelease(openConnection)(c => closeConnection(c).orDie)
        }
        .flatMap(env =>
          ZLayer
            .scoped {
              ZIO
                .acquireReleaseExit(
                  setAutoCommit(env.get, autoCommit = false).as(env.get)
                ) {
                  case (c, Exit.Success(_)) => commitConnection(c).orDie
                  case (c, Exit.Failure(_)) =>
                    (if (commitOnFailure) commitConnection(c)
                     else rollbackConnection(c)).orDie
                }
            }
        )
```

```
      ZStream
        .serviceWithStream[Connection](c => task(c).mapError(Right(_)))
        .provideSomeLayer[R](connectionLayer(commitOnFailure).mapError(Left(_)))
```

Reply:
Hmm, I like the idea of having a single transaction-handler for both the ZIO and ZStream methods :+1:. I want to add an additional distinction though: if the ZIO or ZStream dies (not just an error), we shouldn't try to either commit or rollback. Only closing the connection is the way to go I think.

I'll try and integrate that next week-end.